src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -1723,6 +1749,67 @@ void InterpreterSelectQuery::executeAggregation(QueryPipeline & pipeline, const
     /// Forget about current totals and extremes. They will be calculated again after aggregation if needed.
     pipeline.dropTotalsAndExtremes();
 
+    /// TODO better case determination
+    if (group_by_info && settings.optimize_aggregation_in_order)
+    {
+        auto & query = getSelectQuery();
+        SortDescription group_by_descr = getSortDescriptionFromGroupBy(query, *context);
+        bool need_finish_sorting = (group_by_info->order_key_prefix_descr.size() < group_by_descr.size());
+
+        if (need_finish_sorting)
+        {
```

Comment:
Maybe it would be more efficient to finish aggregation with small hash tables for batches in that case. We can delay this task and try to imlement it later in other PR.